### PR TITLE
[POC] Upgrade es-v8 to OpenSearch java client

### PR DESCRIPTION
**Pull request by Arsnael**
Just a quick POC to check if we can easily upgrade to the java client of OpenSearch: https://opensearch.org/docs/latest/clients/java/

OpenSearch is using two java clients: the legacy high level rest client (forked from es one), and the new java client (base forked from es one before they changed license... so it's very similar to the es v8 java client but with still some differences).

The code is a bit dirty but it's a POC, just to show if it's possible to achieve. I duplicated backends-es and mailbox-es modules to try to do the equivalent with OpenSearch and the java client.

Most of it is green.

I'm not sure it's matured enough for us to switch to it though... A few issues:

- bulk updates with json raw values don't seem to work, thus the dirty workaround in the poc with sending a batch of normal update requests. I opened an issue on their github regarding this
- bug on sorts: `sortOnCcShouldWork` fails because a cc is null on one message, and that case is not taken into consideration. It's a known issue as somebody else opened it on their backlog a month ago (that one is a blocker IMO)
- can't make `searchShouldRetrieveExactlyMatchingMimeMessageID` green: not sure why the search does not return anything here... I don't feel I messed up on the migration (but could have)... but just couldn't put my finger on what's the issue

cc @chibenwa

### CODE REVIEW

File: mailbox/opensearch/src/main/java/org/apache/james/mailbox/opensearch/MailboxMappingFactory.java [hunk 1]
```diff
@@ -19,241 +19,225 @@
...
+                                .build())
+                            .build()
+                    ))
+                    .build())
+                .build())
+            .put(JsonMessageConstants.MAILBOX_ID, new Property.Builder()
+                .keyword(new KeywordProperty.Builder().store(true).build())
+                .build())
+            .put(JsonMessageConstants.MIME_MESSAGE_ID, new Property.Builder()
+                .text(new TextProperty.Builder().analyzer(STANDARD).build())
```
Comment (chibenwa) [inline mailbox/opensearch/src/main/java/org/apache/james/mailbox/opensearch/MailboxMappingFactory.java:206]: Previous analuser was keyword, not standard! That's your issue....

**Comment by chibenwa** [comment]
> bulk updates with json raw values don't seem to work, thus the dirty workaround in the poc with sending a batch of normal update requests. I opened an issue on their github regarding this

Share link?

> bug on sorts: sortOnCcShouldWork fails because a cc is null on one message, and that case is not taken into consideration. It's a known issue as somebody else opened it on their backlog a month ago (that one is a blocker IMO)

Idem.

Also does it work if we replace null with an empty collection?

**Comment by Arsnael** [comment]
> Also does it work if we replace null with an empty collection?

It's null because the eml has no cc (which is a common thing in mails to send some without a cc right?). I guess the listener indexes it and there is no cc in the document headers... but the client then considers it null when doing a sort on it? (cause it doesn't exist) Not sure to get what you want me to check/test here with your empty collection?

**Comment by Arsnael** [comment]
Fix for bulk update and workaround for the sort... I'm done with the POC :)

File: mailbox/opensearch/src/main/java/org/apache/james/mailbox/opensearch/json/EMailers.java [hunk 2]
```diff
@@ -19,13 +19,16 @@
...
+import java.util.Optional;
 import java.util.Set;
 import java.util.stream.Collectors;
 
 import com.fasterxml.jackson.annotation.JsonValue;
 import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableSet;
 
 public class EMailers implements SerializableMessage {
+    public static final EMailers DEFAULT = new EMailers(ImmutableSet.of(new EMailer(Optional.empty(), "zzzzzzzzz")));
```
Comment (vttranlina) [inline mailbox/opensearch/src/main/java/org/apache/james/mailbox/opensearch/json/EMailers.java:31]: "zzzzzzzzz" in the main package?
//  If it is a temp for the test, do not forget to remove it

File: mailbox/opensearch/src/main/java/org/apache/james/mailbox/opensearch/json/IndexableMessage.java [hunk 3]
```diff
@@ -297,6 +297,8 @@ public EMailers getCc() {
 
     @JsonProperty(JsonMessageConstants.CC_SORT)
     public EMailers getCcSort() {
+        // workaround for ISSUE with sorts on null/empty indexed fields:
+        // https://github.com/opensearch-project/opensearch-java/issues/158
```
Comment (chibenwa) [inline mailbox/opensearch/src/main/java/org/apache/james/mailbox/opensearch/json/IndexableMessage.java:301]: Did the fix opened at https://github.com/opensearch-project/opensearch-java/pull/169 fixes this?

**Comment by chibenwa** [comment]
```
<p><span style="font-weight:bold">org.apache.james.webadmin.integration.rabbitmq.RabbitMQWebAdminServerTaskSerializationIntegrationImmutableTest.blobStoreVaultGarbageCollectionShouldComplete</span></p><div style="text-align:right;">Failing for the past 1 build
          (Since <a href="https://ci-builds.apache.org/job/james/job/ApacheJames/job/PR-1049/6/" class="model-link"><span style="width: 16px; height: 16px; " class="build-status-icon__wrapper icon-red icon-sm"><span class="build-status-icon__outer"><svg viewBox="0 0 24 24" focusable="false" class="svg-icon "><use href="https://ci-builds.apache.org/images/build-status/build-status-sprite.svg#build-status-static"></use></svg></span><svg viewBox="0 0 24 24" focusable="false" class="svg-icon icon-red icon-sm"><use href="https://ci-builds.apache.org/static/2864a239/images/build-status/build-status-sprite.svg#last-failed"></use></svg></span><span class="jenkins-icon-adjacent">#6</span></a> )
        </div><div style="text-align:right;"><a href="https://ci-builds.apache.org/job/james/job/ApacheJames/job/PR-1049/6/testReport/junit/org.apache.james.webadmin.integration.rabbitmq/RabbitMQWebAdminServerTaskSerializationIntegrationImmutableTest/blobStoreVaultGarbageCollectionShouldComplete/history">Took 0.23 sec.</a></div><div id="description"><div></div><div class="jenkins-buttons-row jenkins-buttons-row--invert"></div></div>
<h3>Error Message</h3><pre>1 expectation failed.
JSON path status doesn't match.
Expected: is "completed"
  Actual: null
</pre><h3>Stacktrace</h3><pre>java.lang.AssertionError: 
1 expectation failed.
JSON path status doesn't match.
Expected: is "completed"
  Actual: null

	at org.apache.james.webadmin.integration.rabbitmq.RabbitMQWebAdminServerTaskSerializationIntegrationImmutableTest.blobStoreVaultGarbageCollectionShouldComplete(RabbitMQWebAdminServerTaskSerializationIntegrationImmutableTest.java:167)
</pre><h3>Standard Output</h3><pre>13:57:18.067 [ERROR] o.a.j.e.EventBus - Error while calling org.apache.james.task.eventsourcing.TaskExecutionDetailsProjection$$Lambda$2040/0x0000000800d67840@5acc3530 for Created(TaskAggregateId(TaskId{value=c135d9d3-e8ec-4396-96ef-1dc18fa922fa}),EventId(0),org.apache.james.vault.blob.BlobStoreVaultGarbageCollectionTask@26d5308c,Hostname(asf938.gq1.ygridcore.net))
java.lang.NullPointerException: null
	at org.apache.james.vault.blob.BlobStoreVaultGarbageCollectionTaskAdditionalInformationDTO.fromDomainObject(BlobStoreVaultGarbageCollectionTaskAdditionalInformationDTO.java:39)
	at org.apache.james.json.DTOModule.toDTO(DTOModule.java:134)
	at org.apache.james.json.DTOConverter.lambda$toDTO$1(DTOConverter.java:57)
	at java.base/java.util.Optional.map(Optional.java:265)
	at org.apache.james.json.DTOConverter.toDTO(DTOConverter.java:57)
	at org.apache.james.json.JsonGenericSerializer.serialize(JsonGenericSerializer.java:120)
	at org.apache.james.server.task.json.JsonTaskAdditionalInformationSerializer.serialize(JsonTaskAdditionalInformationSerializer.java:61)
	at org.apache.james.task.eventsourcing.cassandra.CassandraTaskExecutionDetailsProjectionDAO.$anonfun$serializeAdditionalInformation$1(CassandraTaskExecutionDetailsProjectionDAO.scala:101)
	at java.base/java.util.Optional.map(Optional.java:265)
	at org.apache.james.task.eventsourcing.cassandra.CassandraTaskExecutionDetailsProjectionDAO.serializeAdditionalInformation(CassandraTaskExecutionDetailsProjectionDAO.scala:101)
	at org.apache.james.task.eventsourcing.cassandra.CassandraTaskExecutionDetailsProjectionDAO.$anonfun$saveDetails$9(CassandraTaskExecutionDetailsProjectionDAO.scala:75)
	at org.apache.james.task.eventsourcing.cassandra.CassandraTaskExecutionDetailsProjectionDAO.$anonfun$saveDetails$10(CassandraTaskExecutionDetailsProjectionDAO.scala:79)
	at scala.collection.LinearSeqOps.foldLeft(LinearSeq.scala:169)
	at scala.collection.LinearSeqOps.foldLeft$(LinearSeq.scala:165)
	at scala.collection.immutable.List.foldLeft(List.scala:79)
	at org.apache.james.task.eventsourcing.cassandra.CassandraTaskExecutionDetailsProjectionDAO.saveDetails(CassandraTaskExecutionDetailsProjectionDAO.scala:78)
	at org.apache.james.task.eventsourcing.cassandra.CassandraTaskExecutionDetailsProjection.update(CassandraTaskExecutionDetailsProjection.scala:39)
	at org.apache.james.task.eventsourcing.TaskExecutionDetailsProjection.$anonfun$asSubscriber$1(TaskExecutionDetailsProjection.scala:32)
	at org.apache.james.eventsourcing.ReactiveSubscriberWrapper.handle(Subscriber.scala:43)
	at org.apache.james.eventsourcing.ReactiveSubscriberWrapper.$anonfun$handleReactive$1(Subscriber.scala:45)
	at org.apache.james.eventsourcing.ReactiveSubscriberWrapper.$anonfun$handleReactive$1$adapted(Subscriber.scala:45)
	at reactor.core.publisher.MonoCallable.call(MonoCallable.java:92)
	at reactor.core.publisher.FluxSubscribeOnCallable$CallableSubscribeOnSubscription.run(FluxSubscribeOnCallable.java:227)
	at reactor.core.scheduler.SchedulerTask.call(SchedulerTask.java:68)
	at reactor.core.scheduler.SchedulerTask.call(SchedulerTask.java:28)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
Request method:	GET
Request URI:	<a href="http://localhost:37739/tasks/c135d9d3-e8ec-4396-96ef-1dc18fa922fa/await">http://localhost:37739/tasks/c135d9d3-e8ec-4396-96ef-1dc18fa922fa/await</a>
Proxy:			&lt;none&gt;
Request params:	&lt;none&gt;
Query params:	&lt;none&gt;
Form params:	&lt;none&gt;
Path params:	&lt;none&gt;
Headers:		Accept=application/json, application/javascript, text/javascript, text/json
				Content-Type=application/json; charset=UTF-8
Cookies:		&lt;none&gt;
Multiparts:		&lt;none&gt;
Body:			&lt;none&gt;

HTTP/1.1 404 Not Found
Date: Thu, 16 Jun 2022 13:57:18 GMT
Access-Control-Allow-Origin: *
Access-Control-Request-Method: DELETE, GET, POST, PUT
Access-Control-Allow-Headers: Content-Type, Authorization, Accept
Content-Type: application/json
Transfer-Encoding: chunked
Server: Jetty(9.4.46.v20220331)

{
    "statusCode": 404,
    "type": "ServerError",
    "message": "The taskId is not found",
    "details": null
}
</pre>org.apache.james.webadmin.integration.rabbitmq.RabbitMQWebAdminServerTaskSerializationIntegrationImmutableTest.blobStoreVaultGarbageCollectionShouldComplete
Failing for the past 1 build (Since
[#6](https://ci-builds.apache.org/job/james/job/ApacheJames/job/PR-1049/6/) )
[Took 0.23 sec.](https://ci-builds.apache.org/job/james/job/ApacheJames/job/PR-1049/6/testReport/junit/org.apache.james.webadmin.integration.rabbitmq/RabbitMQWebAdminServerTaskSerializationIntegrationImmutableTest/blobStoreVaultGarbageCollectionShouldComplete/history)
Error Message

1 expectation failed.
JSON path status doesn't match.
Expected: is "completed"
  Actual: null

Stacktrace

java.lang.AssertionError: 
1 expectation failed.
JSON path status doesn't match.
Expected: is "completed"
  Actual: null

	at org.apache.james.webadmin.integration.rabbitmq.RabbitMQWebAdminServerTaskSerializationIntegrationImmutableTest.blobStoreVaultGarbageCollectionShouldComplete(RabbitMQWebAdminServerTaskSerializationIntegrationImmutableTest.java:167)

Standard Output

13:57:18.067 [ERROR] o.a.j.e.EventBus - Error while calling org.apache.james.task.eventsourcing.TaskExecutionDetailsProjection$$Lambda$2040/0x0000000800d67840@5acc3530 for Created(TaskAggregateId(TaskId{value=c135d9d3-e8ec-4396-96ef-1dc18fa922fa}),EventId(0),org.apache.james.vault.blob.BlobStoreVaultGarbageCollectionTask@26d5308c,Hostname(asf938.gq1.ygridcore.net))
java.lang.NullPointerException: null
	at org.apache.james.vault.blob.BlobStoreVaultGarbageCollectionTaskAdditionalInformationDTO.fromDomainObject(BlobStoreVaultGarbageCollectionTaskAdditionalInformationDTO.java:39)
	at org.apache.james.json.DTOModule.toDTO(DTOModule.java:134)
	at org.apache.james.json.DTOConverter.lambda$toDTO$1(DTOConverter.java:57)
	at java.base/java.util.Optional.map(Optional.java:265)
	at org.apache.james.json.DTOConverter.toDTO(DTOConverter.java:57)
	at org.apache.james.json.JsonGenericSerializer.serialize(JsonGenericSerializer.java:120)
	at org.apache.james.server.task.json.JsonTaskAdditionalInformationSerializer.serialize(JsonTaskAdditionalInformationSerializer.java:61)
	at org.apache.james.task.eventsourcing.cassandra.CassandraTaskExecutionDetailsProjectionDAO.$anonfun$serializeAdditionalInformation$1(CassandraTaskExecutionDetailsProjectionDAO.scala:101)
	at java.base/java.util.Optional.map(Optional.java:265)
	at org.apache.james.task.eventsourcing.cassandra.CassandraTaskExecutionDetailsProjectionDAO.serializeAdditionalInformation(CassandraTaskExecutionDetailsProjectionDAO.scala:101)
	at org.apache.james.task.eventsourcing.cassandra.CassandraTaskExecutionDetailsProjectionDAO.$anonfun$saveDetails$9(CassandraTaskExecutionDetailsProjectionDAO.scala:75)
	at org.apache.james.task.eventsourcing.cassandra.CassandraTaskExecutionDetailsProjectionDAO.$anonfun$saveDetails$10(CassandraTaskExecutionDetailsProjectionDAO.scala:79)
	at scala.collection.LinearSeqOps.foldLeft(LinearSeq.scala:169)
	at scala.collection.LinearSeqOps.foldLeft$(LinearSeq.scala:165)
	at scala.collection.immutable.List.foldLeft(List.scala:79)
	at org.apache.james.task.eventsourcing.cassandra.CassandraTaskExecutionDetailsProjectionDAO.saveDetails(CassandraTaskExecutionDetailsProjectionDAO.scala:78)
	at org.apache.james.task.eventsourcing.cassandra.CassandraTaskExecutionDetailsProjection.update(CassandraTaskExecutionDetailsProjection.scala:39)
	at org.apache.james.task.eventsourcing.TaskExecutionDetailsProjection.$anonfun$asSubscriber$1(TaskExecutionDetailsProjection.scala:32)
	at org.apache.james.eventsourcing.ReactiveSubscriberWrapper.handle(Subscriber.scala:43)
	at org.apache.james.eventsourcing.ReactiveSubscriberWrapper.$anonfun$handleReactive$1(Subscriber.scala:45)
	at org.apache.james.eventsourcing.ReactiveSubscriberWrapper.$anonfun$handleReactive$1$adapted(Subscriber.scala:45)
	at reactor.core.publisher.MonoCallable.call(MonoCallable.java:92)
	at reactor.core.publisher.FluxSubscribeOnCallable$CallableSubscribeOnSubscription.run(FluxSubscribeOnCallable.java:227)
	at reactor.core.scheduler.SchedulerTask.call(SchedulerTask.java:68)
	at reactor.core.scheduler.SchedulerTask.call(SchedulerTask.java:28)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
Request method:	GET
Request URI:	http://localhost:37739/tasks/c135d9d3-e8ec-4396-96ef-1dc18fa922fa/await
Proxy:			<none>
Request params:	<none>
Query params:	<none>
Form params:	<none>
Path params:	<none>
Headers:		Accept=application/json, application/javascript, text/javascript, text/json
				Content-Type=application/json; charset=UTF-8
Cookies:		<none>
Multiparts:		<none>
Body:			<none>

HTTP/1.1 404 Not Found
Date: Thu, 16 Jun 2022 13:57:18 GMT
Access-Control-Allow-Origin: *
Access-Control-Request-Method: DELETE, GET, POST, PUT
Access-Control-Allow-Headers: Content-Type, Authorization, Accept
Content-Type: application/json
Transfer-Encoding: chunked
Server: Jetty(9.4.46.v20220331)

{
    "statusCode": 404,
    "type": "ServerError",
    "message": "The taskId is not found",
    "details": null
}
```

Related ?

**Comment by Arsnael** [comment]
Very unlikely, as the only thing I did was duplicating the modules, they are not wired to the rest of the code

**Comment by chibenwa** [comment]
See https://github.com/apache/james-project/pull/1061

**Pull request closed without merging**